include/hax_interface.h

```diff
@@ -49,6 +49,65 @@
 #include "windows/hax_interface_windows.h"
 #endif
 
+#define HAX_IOCTL_VERSION \
```

Comment:
> **Prefix**: The IOCTLs can be divided into 3 groups (global, VM and vCPU), and they now have different prefixes as well: `HAX_IOCTL`, `HAX_VM_IOCTL`, `HAX_VCPU_IOCTL`. Should we rename the first one to `HAX_GLOBAL`? Or should we use the same prefix for all of them, e.g. just `HAX` (KVM uses this naming scheme)?

Although renaming IOCTLs to have the same "prefix depth", i.e. either `HAX_${action}` (like KVM) or `HAX_${type}_IOCTL_${action}` (like you suggested with *type := {GLOBAL,VM,VCPU}*), seems more consistent, I'm concerned about verbosity vs readability:

- **Verbosity issue**: If we maintain all prefixes and require them to have the same depth, it will cause extremely long names when considering extension ioctl's. Ideally, those kind of ioctl's should have prefixes too, e.g. `HAX_EXT_IOCTL_${action}`. but what should we do with standard global ioctl's then? Calling them e.g. `HAX_STD_GLOBAL_IOCTL_CREATE_VM` seems really verbose (even worse if we add another prefix for platform-specific ioctl's).

- **Readability issue**: If we drop all prefixes except `HAX_`, then we lose valuable information for developers. Prefixes are useful after all.

I think the sweet spot between options is accepting that prefixes might be dropped if not relevant: e.g. global ioctls, standard ioctls, platform-agnostic ioctls. This is what the current prefix scheme already does.

> **Function**: I think this is always an action and should start with a verb. So `VERSION` would be come `GET_VERSION` (or even better, `GET_API_VERSION`).

Fully agreed. This seems also the way KVM is doing it.

---

With these two points in mind. The pseudo-regex/bash code I would propose is:
`HAX_(${platform}_)?(${ext}_)?(${type}_)?IOCTL_${action}`, with the following keywords.
- *${platform}*: Used for platform-specific ioctl codes (otherwise dropped).
- *${ext}*: Written as `EXT_`. Used for extension-specific ioctl codes (otherwise dropped).
- *${type}*: Used for VM/VCPU-specific ioctl codes (otherwise dropped).
- *${action}*: Written as `${verb}_${noun}`.

The way to name platform/extension-specific ioctl's (if any) would be, for example:
- `HAX_EXT_IOCTL_${VERB}_${NOUN}`: Extension ioctl (i.e. non-standard 3rd party ioctl).
- `HAX_WINDOWS_IOCTL_${VERB}_${NOUN}`: Windows-specific ioctl.
- `HAX_DARWIN_EXT_IOCTL_${VERB}_${NOUN}`: MacOS-specific extension ioctl.

The rest of IOCTLs would preserve the current naming, aside from the "verb-less" ones (e.g. *HAX_IOCTL_VERSION*), and some minor inconsistencies in verb placement (e.g. *HAX_IOCTL_CREATE_VM* vs *HAX_VM_IOCTL_VCPU_CREATE*).